app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/repository/RepositoryBrowserViewModel.kt

```diff
@@ -0,0 +1,93 @@
+package com.bottlerocketstudios.brarchitecture.ui.repository
+
+import com.bottlerocketstudios.brarchitecture.R
+import com.bottlerocketstudios.brarchitecture.data.model.ApiResult
+import com.bottlerocketstudios.brarchitecture.data.model.RepoFile
+import com.bottlerocketstudios.brarchitecture.data.repository.BitbucketRepository
+import com.bottlerocketstudios.brarchitecture.infrastructure.toast.Toaster
+import com.bottlerocketstudios.brarchitecture.infrastructure.util.exhaustive
+import com.bottlerocketstudios.brarchitecture.navigation.NavigationEvent
+import com.bottlerocketstudios.brarchitecture.ui.BaseViewModel
+import com.bottlerocketstudios.compose.repository.RepositoryItemUiModel
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.withContext
+
+class RepositoryBrowserViewModel(private val repo: BitbucketRepository, private val toaster: Toaster) : BaseViewModel() {
+
+    private val srcFiles: StateFlow<List<RepoFile>> = MutableStateFlow(emptyList())
+    private var currentRepoName: String = ""
+
+    val path: StateFlow<String> = MutableStateFlow("")
+    val itemCount: StateFlow<Int> = srcFiles
+        .map { it.size }
+        .groundState(0)
+    val uiModels: StateFlow<List<RepositoryItemUiModel>> = srcFiles
+        .map { files ->
+            files.map { file ->
+                RepositoryItemUiModel(
+                    path = file.path ?: "",
+                    size = file.size ?: 0,
+                    isFolder = file.type == "commit_directory"
+                )
+            }
+        }
+        .groundState(emptyList())
+
+    fun getFiles(data: RepositoryBrowserData) {
+        currentRepoName = data.repoName
+        path.set(data.folderPath ?: data.repoName)
+        val selectedRepo = repo.repos.value.firstOrNull { it.name?.equals(data.repoName) ?: false }
+        selectedRepo?.let {
+            val slug = it.workspace?.slug ?: ""
+            val name = it.name ?: ""
+            launchIO {
+                val result = if (data.folderHash != null && data.folderPath != null) {
+                    repo.getSourceFolder(slug, name, data.folderHash, data.folderPath)
+                } else {
+                    repo.getSource(slug, name)
+                }
+                when (result) {
+                    is ApiResult.Success -> srcFiles.set(result.data)
+                    is ApiResult.Failure -> {
+                        // TODO: Improve error messaging
+                        withContext(dispatcherProvider.Main) {
+                            toaster.toast(R.string.error_loading_repository)
+                        }
```

Comment:
I'll be doing some live coding today during Kotlin training this afternoon and creating some standard error handling for this pattern.

Reply:
Unfortunately I had to drop early from the training to get to a client meeting so I missed the payoff on this. :(